zenoh/src/net/routing/interceptor/downsampling.rs

```diff
@@ -0,0 +1,160 @@
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+//! ⚠️ WARNING ⚠️
+//!
+//! This module is intended for Zenoh's internal use.
+//!
+//! [Click here for Zenoh's documentation](../zenoh/index.html)
+
+use crate::net::routing::interceptor::*;
+use std::sync::{Arc, Mutex};
+use zenoh_config::DownsamplingItemConf;
+use zenoh_protocol::core::key_expr::OwnedKeyExpr;
+
+const RATELIMIT_STRATEGY: &str = "ratelimit";
+
+// EgressMsgDownsamplerRatelimit implements default ratelimit strategy, it has following
+// configuration parameters:
+//   threshold  - mandatory, indicates the minimum interval between messages.
+//                All messages arriving with a smaller interval will be skipped
+//   keyexprs   - optional, list of key expressions, only messages matching this expression will be processed,
+//                the rest will be passed as is.
+//   interfaces - optional, list of interfaces messages will be processed on,
+//                the rest will be passed as is.
+//
+// TODO: we can also add decimation strategy, with mandatory "factor" parameter
+
+pub(crate) struct EgressMsgDownsamplerRatelimit {
+    keyexprs: Option<Vec<OwnedKeyExpr>>,
+    threshold: std::time::Duration,
+    latest_message_timestamp: Arc<Mutex<std::time::Instant>>,
+}
+
+impl InterceptorTrait for EgressMsgDownsamplerRatelimit {
+    fn intercept(
+        &self,
+        ctx: RoutingContext<NetworkMessage>,
+    ) -> Option<RoutingContext<NetworkMessage>> {
+        if let Some(cfg_keyexprs) = self.keyexprs.as_ref() {
+            let mut matched = false;
+            if let Some(keyexpr) = ctx.full_key_expr() {
+                for cfg_keyexpr in cfg_keyexprs {
+                    if cfg_keyexpr.intersects(&keyexpr) {
+                        matched = true;
+                    }
+                }
+            }
+            if !matched {
+                return Some(ctx);
+            }
+        }
+
+        let timestamp = std::time::Instant::now();
+        let mut latest_message_timestamp = self.latest_message_timestamp.lock().unwrap();
+
+        if timestamp - *latest_message_timestamp >= self.threshold {
+            *latest_message_timestamp = timestamp;
+            log::debug!("Interceptor: Passed threshold, passing.");
+            Some(ctx)
+        } else {
+            log::debug!("Interceptor: Skipped due to threshold.");
+            None
+        }
+    }
+}
+
+impl EgressMsgDownsamplerRatelimit {
+    pub fn new(conf: DownsamplingItemConf) -> Self {
+        if let Some(threshold_ms) = conf.threshold_ms {
+            let threshold = std::time::Duration::from_millis(threshold_ms);
+            Self {
+                keyexprs: conf.keyexprs,
+                threshold,
+                // TODO (sashacmc): I need just := 0, but how???
+                latest_message_timestamp: Arc::new(Mutex::new(
```

Comment:
I need some value in the past, e.g. zero, but I don't know how better do it in rust.
Any idea?

Reply:
It's not clear to me what you are trying to achieve there... can you provide an example?

Reply:
Reading this: https://github.com/rust-lang/rust/issues/40910, it seems there isn't much better solution using the `Instant` type.

Reply:
@Mallets
I need to filter out the messages that arrive while the time elapsed since the last unfiltered one is less than a specified interval, and I do not want to filter out the first message.

The simplest way (in order not to do unnecessary checks every time) is to set an initial time value that will lag behind the current moment by obviously more than the specified interval.

For C and posix it will looks like
```
latest_timestamp = 0 // == 1970.01.01

void event_hadler(message) {
  timestamp = now();
  if (timestamp - latest_timestamp < threshold) {
    // skip message
  } else {
    latest_timestamp = timestamp;
    // process message
  }  
}
```